include/NAS2D/Renderer/Rectangle.h

```diff
@@ -13,93 +13,109 @@
 namespace NAS2D {
 
 
-struct Rectangle_2d;
-struct Rectangle_2df;
+template <typename BaseType>
+struct Rectangle
+{
+	Rectangle() = default;
+	Rectangle(const Rectangle& other) = default;
+	Rectangle& operator=(const Rectangle& other) = default;
+
```

Comment:
Consider `=default`ing the move operations and the Dtor to adhere to Rule of Five/Six and [C.21](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-five).

It's not necessarily a requirement (and I'm still not sure if moves of POD types aren't silently converted to copies anyway) but may be a missed opportunity for compiler optimization.

Reply:
Well, that got me thinking.

It doesn't seem to make much sense to move a POD type. Really what you'd want is to just copy it. There's no need to move from it, and update where it comes from. Effectively, there is no external resource being managed here. That's is largely why the special operations are all defaulted, rather than being user provided. The only reason they were explicitly defaulted, is because the constructor which takes initial values would have caused them to be deleted. And that only exists because we haven't yet transitioned this fully from a `class` to a `struct`, and we still have private fields.

According to the [The rule of three/five/zero](https://en.cppreference.com/w/cpp/language/rule_of_three), it talks about user defined special operations. None of those operations are user defined, only defaulted. It also talks about managing resources, which we don't do.

I would say the rule of three/five doesn't apply here.

----

We'll probably need to revisit the special member functions once we compete the transition and make the member fields public. My guess is, explicit mention of the special member functions will all disappear, along with the constructor taking initial values. The compiler should provide appropriate defaults at that time.

Reply:
If we can get rid of *all* the default operations, i.e., the rule of zero, that would be even better! :)